internal/telemetry/metrics/providers.go

```diff
@@ -73,32 +79,60 @@ func registerDefaultViews() error {
 
 // newProxyMetricsHandler creates a subrequest to the envoy control plane for metrics and
 // combines them with our own
-func newProxyMetricsHandler(promHandler http.Handler, envoyURL url.URL) http.HandlerFunc {
+func newProxyMetricsHandler(exporter *ocprom.Exporter, envoyURL url.URL, installationID string) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		defer promHandler.ServeHTTP(w, r)
-
 		// Ensure we don't get entangled with compression from ocprom
 		r.Header.Del("Accept-Encoding")
 
-		r, err := http.NewRequestWithContext(r.Context(), "GET", envoyURL.String(), nil)
+		rec := httptest.NewRecorder()
+		exporter.ServeHTTP(rec, r)
+
+		err := writeMetricsWithInstallationID(w, rec.Body, installationID)
+		if err != nil {
+			log.Error().Err(err).Send()
+			return
+		}
+
+		req, err := http.NewRequestWithContext(r.Context(), "GET", envoyURL.String(), nil)
 		if err != nil {
 			log.Error().Err(err).Msg("telemetry/metrics: failed to create request for envoy")
 			return
 		}
 
-		resp, err := http.DefaultClient.Do(r)
+		resp, err := http.DefaultClient.Do(req)
 		if err != nil {
 			log.Error().Err(err).Msg("telemetry/metrics: fail to fetch proxy metrics")
 			return
 		}
 		defer resp.Body.Close()
 
-		envoyBody, err := ioutil.ReadAll(resp.Body)
+		err = writeMetricsWithInstallationID(w, resp.Body, installationID)
 		if err != nil {
-			log.Error().Err(err).Msg("telemetry/metric: failed to read proxy metrics")
+			log.Error().Err(err).Send()
 			return
 		}
+	}
+}
 
-		w.Write(envoyBody)
+func writeMetricsWithInstallationID(w io.Writer, r io.Reader, installationID string) error {
+	var parser expfmt.TextParser
+	ms, err := parser.TextToMetricFamilies(r)
+	if err != nil {
+		return fmt.Errorf("telemetry/metric: failed to read prometheus metrics: %w", err)
+	}
+
+	for _, m := range ms {
+		for _, mm := range m.Metric {
```

Comment:
Any potential performance concerns of this being O(M*N)? ; I don't have a good sense of the potential dimensionality of the `TextToMetricFamilies` struct.

Reply:
Each metric is a separate line. Not sure how else we could solve this. I didn't find anything in the envoy docs about how to add arbitrary tags to metrics.